test/jdk/javax/swing/JTextArea/4514331/TabShiftsFocusToNextComponent.java

```diff
@@ -0,0 +1,134 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.awt.Point;
+import java.awt.Robot;
+import java.awt.event.FocusAdapter;
+import java.awt.event.FocusEvent;
+import java.awt.event.InputEvent;
+import java.awt.event.KeyEvent;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.swing.JButton;
+import javax.swing.JFrame;
+import javax.swing.JPanel;
+import javax.swing.JTextArea;
+import javax.swing.SwingUtilities;
+import javax.swing.UIManager;
+
+import static javax.swing.UIManager.getInstalledLookAndFeels;
+
+/*
+ * @test
+ * @key headful
+ * @bug 4514331
+ * @summary Check whether pressing <Tab> key always shift focus to next component,
+ *          even though the current focus is in JTextArea and some text is already selected.
+ * @run main TabShiftsFocusToNextComponent
+ */
+public class TabShiftsFocusToNextComponent {
+
+    private static JFrame frame;
+    private static JTextArea editor;
+    private static Robot robot;
+    private static boolean passed = false;
+
+    public static void main(String[] s) throws Exception {
+        runTest();
+    }
+
+    public static void runTest() throws Exception {
+        robot = new Robot();
+        robot.setAutoWaitForIdle(true);
+        robot.setAutoDelay(200);
+        List<String> lafs = Arrays.stream(getInstalledLookAndFeels())
+                                  .map(UIManager.LookAndFeelInfo::getClassName)
+                                  .collect(Collectors.toList());
+        for (final String laf : lafs) {
+            try {
+                SwingUtilities.invokeAndWait(() -> {
+                    setLookAndFeel(laf);
+                    createUI();
+                });
+                passed = false;
+                Point editorLoc = editor.getLocationOnScreen();
+                robot.mouseMove(editorLoc.x, editorLoc.y);
+                robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+                robot.mouseMove(editorLoc.x + 20, editorLoc.y);
+                robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+                robot.keyPress(KeyEvent.VK_TAB);
+                robot.keyRelease(KeyEvent.VK_TAB);
+                if (passed) {
+                    System.out.println(" Test passed for " + laf);
+                } else {
+                    throw new RuntimeException("Test failed for " + laf);
+                }
+            } finally {
+                SwingUtilities.invokeAndWait(TabShiftsFocusToNextComponent::disposeFrame);
+            }
+        }
+    }
+
+
+    private static void createUI() {
+        frame = new JFrame();
+        JPanel panel = new JPanel();
+        editor = new JTextArea("I am a JTextArea");
+        editor.setEditable(false);
+        editor.requestFocusInWindow();
+        panel.add(editor);
```

Comment:
I am not sure that this will work all the time, the requestFocusInWindow(), see part of the spec:
>  * This component must be displayable, focusable, visible
    * and all of its ancestors (with the exception of the top-level
    * Window) must be visible for the request to be granted

Reply:
I have moved the editor.requestFocusInWindow() call after frame.setVisible(true) and also even if this call is not honoured, the editor will get the focus as we are clicking on the editor (robot.mousePress) in runTest().

Reply:
Made changes for this.